### PR TITLE
Fix the execution hangs due to the infinite loop when handling BuiltI…

### PR DESCRIPTION
**Pull request by ruiminzhao**
…nSampleMask in patchFsBuiltInInputImport()

There is a concer case in a application scenario:
1. RasterizerState.pixelShaderSamples is 0 because it was initialized in GraphicsPipeline::Create() with “GraphicsPipelineBinaryCreateInfo binaryCreateInfo = {};” and not modified after in this case.
2. When it is 0, “baseMaskSamples *= 2” doesn’t change its value so it is stuck in the while loop forever

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9cf97a74576beefd52e4d9cc2a25b0e27fa725b6
<details><summary>CTS tests (Failed: 0/137941)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35352/68939 (51.3%)</li>
<li>Failed: 0/68939 (0.0%)</li>
<li>Not Supported: 33587/68939 (48.7%)</li><li>Warnings: 0/68939 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35424/69002 (51.3%)</li>
<li>Failed: 0/69002 (0.0%)</li>
<li>Not Supported: 33578/69002 (48.7%)</li><li>Warnings: 0/69002 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchInOutImportExport.cpp [hunk 1]
```diff
@@ -2483,7 +2483,9 @@ Value *PatchInOutImportExport::patchFsBuiltInInputImport(Type *inputTy, unsigned
                                             {ancillary, builder.getInt32(8), builder.getInt32(4)});
 
     Value *sampleMaskIn = sampleCoverage;
-    if (m_pipelineState->getRasterizerState().perSampleShading || builtInUsage.runAtSampleRate) {
+    // It should keep pixelShaderSamples be valid value, otherwise it will hang in while loop.
```
Comment (nhaehnle) [inline lgc/patch/PatchInOutImportExport.cpp:2486]: This comment is not a good justification for why this change ought to be correct.

What is the justification for why this is correct? When can it happen that pixelShaderSamples == 0 but runAtSampleRate is true?

**Review by nhaehnle — COMMENTED** [review]
Thanks for adding the test.

File: lgc/patch/PatchInOutImportExport.cpp (hunk 1 above)
Comment (ruiminzhao) [reply lgc/patch/PatchInOutImportExport.cpp:2486]: Thanks for the comments. I have some investigation before:

From some comments in code:
```
runAtSampleRate: Whether fragment shader run at sample rate
pixelShaderSamples:  Controls the pixel shader execution rate. Must be less than or equal to coverageSamples. Valid values are 1, 2, 4, and 8.
```
The pixelShaderSamples will be set from driver when BuildMultisampleState, seems the value for this will be set from API.
But the runAtSampleRate will be set to 1 when read builtInvariable(BuiltInSampleId/BuiltInSamplePosition/..) or have InterpLocSample infos, which is set when compile shader.

So in next product, broadcast samplemask will only be used when pixelShaderSamples is set.

I will expand the comments.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 27f8e38f9cfb599c1a9a38eb0dda4c536cf0168b
<details><summary>CTS tests (Failed: 0/138020)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35401/69018 (51.3%)</li>
<li>Failed: 0/69018 (0.0%)</li>
<li>Not Supported: 33617/69018 (48.7%)</li><li>Warnings: 0/69018 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35424/69002 (51.3%)</li>
<li>Failed: 0/69002 (0.0%)</li>
<li>Not Supported: 33578/69002 (48.7%)</li><li>Warnings: 0/69002 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
Thanks. Why is skipping the sample coverage masking really the right thing to do in this case? If the PS ends up running at sample rate, wouldn't it make sense to apply the masking with `pixelShaderRate == 1`?

**Comment by ruiminzhao** [comment]
> Thanks. Why is skipping the sample coverage masking really the right thing to do in this case? If the PS ends up running at sample rate, wouldn't it make sense to apply the masking with `pixelShaderRate == 1`?

Oh....my mistake here. Now I have updated the code. 
Thanks for your correction!

**Comment by amdvlk-admin** [comment]
### Test summary for commit 1104f712b53656fa80d7f9f69e9002f368f64f52
<details><summary>CTS tests (Failed: 0/137941)</summary><ul><li>Built with version 1.3.5.2</li>
<details open><summary>Ubuntu navi3x, Srdcvk</summary><ul>

<li>Passed: 35353/68939 (51.3%)</li>
<li>Failed: 0/68939 (0.0%)</li>
<li>Not Supported: 33586/68939 (48.7%)</li><li>Warnings: 0/68939 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu navi2x, Srdcvk</summary><ul>

<li>Passed: 35424/69002 (51.3%)</li>
<li>Failed: 0/69002 (0.0%)</li>
<li>Not Supported: 33578/69002 (48.7%)</li><li>Warnings: 0/69002 (0.0%)</li>
</ul></details>


</ul></details>
